src/partr.c

```diff
@@ -358,7 +355,7 @@ JL_DLLEXPORT void jl_wakeup_thread(int16_t tid)
     JULIA_DEBUG_SLEEPWAKE( wakeup_enter = cycleclock() );
     if (tid == self || tid == -1) {
         // we're already awake, but make sure we'll exit uv_run
-        if (ptls->sleep_check_state != not_sleeping)
+        if (jl_atomic_load_relaxed(&ptls->sleep_check_state) == sleeping)
             jl_atomic_store(&ptls->sleep_check_state, not_sleeping);
```

Comment:
So this pair of operations looks like a CAS, but done non-atomically. Can that be a problem?

Reply:
It's not a data race, because the only other writer is monotonic (makes the same change). In fact, we can probably relax the atomic store here.

Reply:
Right I think that makes sense. I wondered whether the store could be less strict.

Reply:
Upon further reflection the atomic does guard the memory contents of the work queue, so we might want something stronger. I need to think about it.

Reply:
Yeah, this not reads like an CAS, but it's not. I'd suggest writing this as `!= sleeping`, so it's clearer that it's just a regular (relaxed) store.

Reply:
I prefer the version which avoids the double negative. To be more explanatory it would be helpful to factor the uses of `sleep_check_state` into a minimum of locations and expand the comments describing which state/invariant it's meant to protect and how this implies the atomic ops which are used.